File: mcp/crm-server.ts
/**
 * CRM MCP server — Lab S2.4.
 *
 * Exposes one read-only tool, `list_accounts`, backed by the same
 * synthetic seed data as the Slice 1 API (`seeds/accounts.json`).
 * No write tools. No forbidden fields. Stable empty input schema.
 *
 * Run directly:   npx tsx mcp/crm-server.ts
 * Register with Claude Code (project-scoped, committed to .mcp.json):
 *   claude mcp add --scope project crm -- npx tsx mcp/crm-server.ts
 */

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const seedsPath = join(__dirname, "..", "seeds", "accounts.json");

interface Contact {
  id: string;
  accountId: string;
  name: string;
  role: string;
  email: string;
}

interface Account {
  id: string;
  name: string;
  industry: string;
  renewalDate: string | null;
  contacts: Contact[];
}

const ALLOWED_ACCOUNT_FIELDS = [
  "id",
  "name",
  "industry",
  "renewalDate",
  "contacts",
] as const;

function loadAccounts(): Account[] {
  const raw = JSON.parse(readFileSync(seedsPath, "utf-8")) as unknown[];
  if (!Array.isArray(raw)) {
    throw new Error("seeds/accounts.json is not an array");
  }
  return raw.map((entry) => {
    const account = entry as Record<string, unknown>;
    const projected: Record<string, unknown> = {};
    for (const field of ALLOWED_ACCOUNT_FIELDS) {
      projected[field] = account[field];
    }
    if (!Array.isArray(projected.contacts)) {
      projected.contacts = [];
    }
    return projected as unknown as Account;
  });
}

const server = new McpServer({
  name: "crm",
  version: "0.1.0",
});

server.registerTool(
  "list_accounts",
  {
    title: "List CRM accounts",
    description:
      "Returns all synthetic CRM accounts with embedded contacts, including renewalDate so renewal questions can be answered. Read-only; no arguments.",
    inputSchema: {},
  },
  async () => {
    const accounts = loadAccounts();
    return {
      content: [
        {
          type: "text",
          text: JSON.stringify(accounts, null, 2),
        },
      ],
      structuredContent: { accounts },
    };
  },
);

async function main(): Promise<void> {
  const transport = new StdioServerTransport();
  await server.connect(transport);
}

main().catch((error) => {
  // eslint-disable-next-line no-console
  console.error("crm MCP server failed:", error);
  process.exit(1);
});

File: mcp/crm_server.py
"""CRM MCP server — Lab S2.4 second-language sketch (Python port).

Mirrors the TypeScript canonical at ``mcp/crm-server.ts``: one read-only
stdio tool ``list_accounts``, empty input schema, seed data projected
through an allow-list so no forbidden fields leak. No write surface.

Not executed in the lab. To run locally a reader would:
    uv add "mcp[cli]>=1.2"
    uv run python mcp/crm_server.py
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from mcp.server.fastmcp import FastMCP

SEEDS_PATH = Path(__file__).resolve().parent.parent / "seeds" / "accounts.json"

# Defensive allow-list — mirrors ALLOWED_ACCOUNT_FIELDS in crm-server.ts.
# Anything in the seed file outside this set is stripped before it leaves
# the process, so a future seed edit cannot accidentally widen the surface.
ALLOWED_ACCOUNT_FIELDS: tuple[str, ...] = (
    "id",
    "name",
    "industry",
    "renewalDate",
    "contacts",
)


def load_accounts() -> list[dict[str, Any]]:
    raw = json.loads(SEEDS_PATH.read_text(encoding="utf-8"))
    if not isinstance(raw, list):
        raise ValueError("seeds/accounts.json is not an array")

    projected: list[dict[str, Any]] = []
    for entry in raw:
        account = {field: entry.get(field) for field in ALLOWED_ACCOUNT_FIELDS}
        # Empty-contacts-as-[] contract — matches the TS server and the
        # Slice 1 REST API: never null, always an array.
        if not isinstance(account.get("contacts"), list):
            account["contacts"] = []
        projected.append(account)
    return projected


mcp = FastMCP("crm")


@mcp.tool(
    name="list_accounts",
    title="List CRM accounts",
    description=(
        "Returns all synthetic CRM accounts with embedded contacts, including "
        "renewalDate so renewal questions can be answered. Read-only; no arguments."
    ),
)
def list_accounts() -> dict[str, list[dict[str, Any]]]:
    """No parameters → empty input schema, matching the TS server."""
    return {"accounts": load_accounts()}


if __name__ == "__main__":
    # FastMCP.run() defaults to stdio transport, equivalent to wiring
    # StdioServerTransport explicitly in the TS server.
    mcp.run()
